ui/dashboards/src/views/ViewDashboard/ViewDashboard.tsx

```diff
@@ -53,6 +56,12 @@ export function ViewDashboard(props: ViewDashboardProps) {
   const dashboardRefreshInterval = spec.refreshInterval ?? DEFAULT_REFRESH_INTERVAL;
   const initialTimeRange = useInitialTimeRange(dashboardDuration);
   const initialRefreshInterval = useInitialRefreshInterval(dashboardRefreshInterval);
+  const builtinVariableDefinitions = getBuiltinVariableDefinitions(dashboardResource);
```

Comment:
If you do that, it means that builtin variables from/to won´t be available in a panel embedded use case. 

@sjcobb, is this a use case of Perses to embed only a panel without dashboard? What would be then the builtin variables expected in that case? 

I'm not sure also about the embedded dashboard use case. I guess the builtin variable project will not be valuable in that case?

Reply:
Builtin variables definitions should be injected the same way dashboard variables definitions are injected. And I think it the case with current implementation, no?

Reply:
Yeah you’re right, embedded dashboards and panels would want these, it should be more ingrained into the plugin-system instead of using the similar approach to externalVariableDefinitions

Reply:
I also wonder if plugins should be able to opt-out somehow if project doesn’t apply to them (here is datasource specific [built-in ex](https://github.com/perses/perses/blob/3a637aae1eafcd3d24e4f8ff7a3c5e8b6c971d9e/ui/prometheus-plugin/src/plugins/prometheus-time-series-query/get-time-series-data.ts#L51), no totally relevant here but in case you haven’t seen how that works)

Reply:
I didn't know about this `$__rate_interval`, I will take a look

Reply:
some will have to be ingrained (from, to, dashboard?), some not (project).

Reply:
This is the complexity of this feature :smile: we have similar problems as with var def / external var def, but now all at once.

The difference is that now, there's no action coming from user, so action to change definitions. But still probably actions to change some state. 

We have to decide if we want to create other variables / state in the store, or create a second store/provider etc... To me put them in the same store makes sense as it is still variables and it will help you to inject them in the different panels. But definitely I'm not fan neither of reusing externalVariableDefinitions.


Also you have several layers now to manage: 
- $__project ............. will be available in Perses App
- $__dashboard ...... will be available in Perses App + embedded dashboard
- $__from & $__to ... will be available in Perses App + embedded dashboard + embedded panels

Reply:
Yeah I don't have a great solution for app-specific variables, but `$__from` and `$__to` should be straightforward, some quick pseudo-code: 

```typescript
const { absoluteTimeRange } = useTimeRange();
replace('$__from', absoluteTimeRange.start); 
replace('$__to', absoluteTimeRange.end); 
```

For app-specific, I wonder if we could use a similar approach the [our chartsTheme](https://github.com/perses/perses/blob/3a637aae1eafcd3d24e4f8ff7a3c5e8b6c971d9e/ui/app/src/context/DarkMode.tsx#L50-L59). In that case, we wanted ability to let apps use a different ECharts theme or define different default threshold colors, opt out of tooltip pinning, etc.

How `defaultPluginKinds` is implemented could also be helpful to reference

Reply:
ah okay so you wouldn't leverage at all the templateVariableProvider but just the time range hook implementation.

Note that we have also in a second phase, the datasource-specific builtin variable :partying_face:

Reply:
Aaaah I missed some messages okay nevermind you already talekd about that

Reply:
So I think that `replace` could live in plugin-system (possibly replace spec.query before it ever gets passed to prometheus-plugin?) 

A couple old PRs from Shan too that could be good to look at: [1](https://github.com/perses/perses/pull/1225/files#diff-52838fcf6ba629b1e359aae1ed07659b1c8165e64172d803937dcc96937bc0f4), [2](https://github.com/perses/perses/pull/1233/files#diff-fd8b76d3a0b28ddc83e9fc2f372e859de53cbce7e47b6626ab46b4f08f428fd0)
And his [comment here](https://github.com/perses/perses/discussions/226#discussioncomment-6092683) is important, we could have regex checking for `__` and would need to add validation in user-defined that does not allow users to name a variable with the `__` prefix. If we had that standardized could even build it directly into [useReplaceVariablesInString](https://github.com/perses/perses/blob/3a637aae1eafcd3d24e4f8ff7a3c5e8b6c971d9e/ui/plugin-system/src/runtime/template-variables.ts#L161C10-L167) 🤔

Reply:
This repo is out of date now but to understand embed-use case it can be helpful to look at, [here are where providers are defined](https://github.com/sjcobb/perses-embed-example/blob/eeefcb48451faab301f88f6341031e645d29dc52/components/dashboards/PersesDashboardProvidersAlt.tsx#L46-L49) and that app can decide to change 'TimeSeriesChart' as the default panel kind if they wanted to.

And here is example of how an app can [override the default ECharts theme](https://github.com/sjcobb/perses-embed-example/blob/eeefcb48451faab301f88f6341031e645d29dc52/components/dashboards/PersesDashboardProviders.tsx#L19-L42C14). I'm thinking that the new provider @celian-garcia mentioned will be needed and it can inject app-specific context about built-in variables. Then TemplateVariableProvider can use the injected dependencies without caring about where they come from (so if that demo Next.js app has no concept of "projects" it could inject info about "workspaces" or "collections" or if RedHat / OpenShift has some other grouping concept in their info model)

Reply:
Or hopefully the Prometheus UI someday will use Perses 🤞 and could leverage this new provider. In that case, they'll be using a panel to replace [the UI here](https://prometheus.demo.do.prometheus.io/graph?g0.expr=up&g0.tab=0&g0.stacked=0&g0.range_input=1h)

Reply:
Thanks for the examples, this will be very useful as we're currently thinking to integrate a perses panel in our home made alerting configuration tool in Amadeus. :wink: